### PR TITLE
Add Prometheus-readable metrics endpoint

### PR DESCRIPTION
**Pull request by DiegoDiverio**
Fixes part of https://github.com/channable/hoff/issues/99 by showing Prometheus-readable metrics on a configurable port, only showing one metric (number of PRs merged) at the moment. Other metrics will be easier to add now, but they should be described in new issues, in more specific detail, when the need arises.

### CODE REVIEW

**Review by bertptrs — COMMENTED** [review]
I don't read Haskell very well so I can't comment too much, but I have one suggestion

Right now we have a single counter for the number of PRs merged. This is not that helpful if you have many projects in a single instance like we do. Perhaps we could add a `project` label to the counter and track the metric by project?

File: app/Main.hs [hunk 1]
```diff
@@ -162,24 +162,83 @@ runMain options = do
...
+  void $ Async.waitAny $ [serverThread, ghThread] ++ metricsThread ++ projectThreads
+
+data ProjectThreadData = ProjectThreadData
+  { projectThreadConfig   :: Config.ProjectConfiguration
+  , projectThreadQueue    :: Logic.EventQueue
+  , projectThreadStateVar :: Logic.StateVar
+  , projectThreadState    :: ProjectState
+  }
+
+projectThread :: Configuration
```
Comment (FPtje) [inline app/Main.hs:205]: Thanks for splitting up this function from the main function. That makes the code more readable!

File: app/Main.hs [hunk 2]
```diff
@@ -162,24 +162,83 @@ runMain options = do
...
     -- sync with the current state of GitHub, accounting for any webhooks that
     -- we missed while not running, or just to fill the state initially after
     -- setting up a new project.
     liftIO $ Logic.enqueueEvent projectQueue Logic.Synchronize
-
-    let
+    -- Start a worker thread to run the main event loop for the project.
+    Async.async
+      $ void
+      -- Implement a newtype for a logging monad that also embodies MonadMonitor
```
Comment (FPtje) [inline app/Main.hs:219]: I'm not sure I understand this comment. This code does not implement a newtype. That is done in `Metrics/Metrics.hs`. Two lines below, the `Metrics.runLoggingMonitorT` does force the monad to be a `LoggingMonitorT`

**Comment by DiegoDiverio** [comment]
> I don't read Haskell very well so I can't comment too much, but I have one suggestion
> 
> Right now we have a single counter for the number of PRs merged. This is not that helpful if you have many projects in a single instance like we do. Perhaps we could add a `project` label to the counter and track the metric by project?

Sorry for the lack of clarity, that's what it already does!

**Review by FPtje — CHANGES_REQUESTED** [review]
Thanks for the PR! This is going to give some wonderful insights!

Some comments, and a general note:

Could you test this PR and make sure:

- The counter is indeed increased when a PR is merged
- The /metrics endpoint gives a good output (you could post it in this PR)

File: src/Configuration.hs [hunk 3]
```diff
@@ -62,6 +64,13 @@ data TlsConfiguration = TlsConfiguration
   }
   deriving (Generic, Show)
 
+data MetricsConfiguration = MetricsConfiguration
+  {
+    metricsPort :: Warp.Port,
+    metricsHost :: String
```
Comment (FPtje) [inline src/Configuration.hs:70]: For consistency this should be a `Text`.

File: src/Metrics/Metrics.hs [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+import Prometheus
+import Control.Monad.IO.Class (MonadIO (liftIO))
+import Control.Monad.Free (Free)
+import Control.Monad.Logger (LoggingT, MonadLogger, NoLoggingT)
+import Control.Monad.Free.Ap (liftF)
+
+type ProjectLabel = Text
+
+data ProjectMetrics = ProjectMetrics
+  { projectMetricsMergedPR     :: Vector ProjectLabel Counter
```
Comment (FPtje) [inline src/Metrics/Metrics.hs:28]: Nit: The indentation of the `::` seems arbitrary here

**Review by FPtje — APPROVED** [review]
Awesome, one nit but otherwise LGTM!

**Review by FPtje — APPROVED** [review]
Awesome!

**Comment by DiegoDiverio** [comment]
@OpsBotPrime merge

**Comment by OpsBotPrime** [comment]
Pull request approved for merge by @DiegoDiverio, rebasing now.

**Comment by OpsBotPrime** [comment]
[CI job :yellow_circle:](https://channable.semaphoreci.com/workflows/a16923f0-b2cd-4983-8f0a-143f2e2dbde4?pipeline_id=e8f29855-ef34-4bfa-a484-7a9a4d8ba0b9) started.

**Pull request merged**